tools/testing/tests/fido2.py

```diff
@@ -33,6 +34,66 @@ def VerifyAttestation(attest, data):
     verifier().verify(attest.att_statement, attest.auth_data, data.hash)
 
 
+def TestCborKeysSorted(cbor_bytes):
+    # Cbor canonical ordering of keys.
+    # https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#ctap2-canonical-cbor-encoding-form
+    def sort_cbor_keys(x1, x2):
```

Comment:
(Not totally sure that I grasp the context here.)

CTAP2 only uses integers and strings as keys, although obviously there's no problem with having a more general function. But this code seems to try comparing the contents of bytestrings and strings, but bytestrings always come first because of the lower major type. Also, the ordering of negative number is tricky because they're [inverted in CBOR](https://tools.ietf.org/html/rfc7049#section-2.1). Thus -2 < -3.

So I would imagine something like:

```python
def cbor_to_representative(key):
  if isinstance(key, int):
    if key >= 0:
      return (0, key)
    return (1, -key)
  elif isintance(key, bytes):
    return (2, key)
  elif isinstance(key, str):
    return (3, key)
  else:
    # any other types that you use?
    raise ValueError(key)

def cmp_cbor_keys(a, b):
  return cmp(cbor_to_representative(a), cbor_to_representative(b))
```

(I don't write much Python; that probably doesn't work.)